File: tp01comentados/PalindromoRecursivo.java


import java.io.*;
import java.nio.charset.*;

class MyIO {

   private static BufferedReader in = new BufferedReader(new InputStreamReader(System.in, Charset.forName("ISO-8859-1")));
   private static String charset = "ISO-8859-1";

   public static void setCharset(String charset_){
      charset = charset_;
      in = new BufferedReader(new InputStreamReader(System.in, Charset.forName(charset)));
   }
    public static String readLine(){
      String s = "";
      char tmp;
      try{
         do{
            tmp = (char)in.read();
            if(tmp != '\n' && tmp != 13){
               s += tmp;
            }
         }while(tmp != '\n');
      }catch(IOException ioe){
         System.out.println("lerPalavra: " + ioe.getMessage());
      }
      return s;
   }
    public static void println(String x){
      try {
         PrintStream out = new PrintStream(System.out, true, charset);
         out.println(x);
      }catch(UnsupportedEncodingException e){ System.out.println("Erro: charset invalido"); }
   }
}
class PalindromoRecursivo
{
//funcao que inicializa a recursividade   
    static boolean ispalin(String s)
    {
	    return ispalin(s, 0);
    }
//funcao que recebe como parametro uma string e um inteiro e retorna um booleano true caso a string seja um palindromo
    public static boolean ispalin(String s, int i) 
    {    
	 int tam = s.length()-1;
//condicao de parada	 
		 if(i==tam)
	 {
		 return true;
	 }	 
         if( s.charAt(i) == s.charAt(tam-i))
	 {
		return true && ispalin(s, i+1);
	 }
	 else 
		 return false; 
          }
    public static void main (String[] args)
    {
      boolean status = true;
// while com condicao = true ate que o if com um break tenha sua condicao atendida
      while (status)
      {
	      String s = MyIO.readLine();
	      if(s.charAt(0) == 'F' && s.charAt(1) == 'I' && s.charAt(2) == 'M' && s.length()== 3)
	      {
		      break;
	      }
         
              if(ispalin(s) == true)
              {
               MyIO.println("SIM");
              }
              else
              {
              MyIO.println("NAO"); 
              }
       }  
     }
}

File: tp01comentados/PalindromoRecursivo.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <stdbool.h>
/* funcao que recebe como parametro uma string e um inteiro e retorna true caso a string seja um palindromo
*/
bool ispalin(char s[500], int i)
{
      int tam = strlen(s)-1;
//condicao de parada
      if(i==tam)
      {
          return true;
      }
        if (s[i]== s[tam-i])
        {
            return true && ispalin(s, i+1);
        }
    else
        return false;
}

int main()
{
    bool status = true;
// while com condicao = true ate que o if com break tenha sua condicao atendida
    while(status)
    {
    char string[500];
    scanf("%[^\n\r]", string);
    while (getchar() != '\n');
    if(string[0]== 'F' && string[1]== 'I' && string[2]== 'M' && strlen(string)== 3)
    {
        break;
    }
//imprimindo o resultado
    bool resp =ispalin(string, 0);
    if(resp == true)
    {
        printf("SIM\n");
    }
    else
        printf("NAO\n");

    }
    return 0;
}
